File: services/staff/control-center/src/features/files/vfs.ts
import { requestSignal } from "@/shared/api/client";
import {
  listVfsNodes,
  searchVfs,
} from "@/shared/api/generated/openapi/sdk.gen";
import type {
  SearchVfsData,
  VfsNode,
  VfsNodePage,
} from "@/shared/api/generated/openapi/types.gen";
import { unwrap } from "@/shared/api/problem";

export async function loadVfsPage(options: {
  path: string;
  query: string;
  projectRef?: string;
  pageToken?: string;
  lifecycleState?: SearchVfsData["query"]["lifecycleState"];
  kinds?: SearchVfsData["query"]["kinds"];
  signal: AbortSignal;
}): Promise<VfsNodePage> {
  const { projectRef, pageToken } = options;
  const signal = AbortSignal.any([options.signal, requestSignal()]);
  const query = options.query.trim();
  const pagination = {
    projectRef,
    pageToken,
    pageSize: 30,
    path: options.path,
    ...(options.lifecycleState
      ? { lifecycleState: options.lifecycleState }
      : {}),
    ...(options.kinds?.length ? { kinds: options.kinds } : {}),
  };
  const page = (
    await unwrap(
      query
        ? searchVfs({ query: { ...pagination, query }, signal })
        : listVfsNodes({
            query: pagination,
            signal,
          }),
    )
  ).data;
  validateVfsPage(page, options);
  return page;
}

export function validateVfsPage(
  page: VfsNodePage,
  options: { projectRef?: string; path: string; query: string },
): void {
  if (
    !Array.isArray(page.items) ||
    !Number.isSafeInteger(page.total) ||
    page.total < 0 ||
    typeof page.nextPageToken !== "string" ||
    page.nextPageToken.length > 2048
  )
    throw new Error("Invalid VFS page");
  const refs = new Set<string>();
  for (const node of page.items) {
    if (
      !(node as VfsNode | null) ||
      typeof node.ref !== "string" ||
      !node.ref ||
      refs.has(node.ref) ||
      typeof node.name !== "string" ||
      !node.name ||
      typeof node.path !== "string" ||
      !node.path.startsWith("/projects/") ||
      typeof node.parentPath !== "string" ||
      typeof node.directory !== "boolean" ||
      ![
        "DIRECTORY",
        "PROJECT",
        "AGENT",
        "WORKFLOW",
        "RUN",
        "INPUT",
        "RESULT",
        "SKILL",
        "MEMORY",
        "AUTOMATION",
        "ENVIRONMENT",
        "AVATAR",
      ].includes(node.kind) ||
      !Number.isSafeInteger(node.sizeBytes) ||
      node.sizeBytes < 0 ||
      typeof node.projectRef !== "string" ||
      !node.projectRef ||
      typeof node.entityRef !== "string" ||
      typeof node.runRef !== "string" ||
      typeof node.digest !== "string" ||
      !Number.isSafeInteger(node.version) ||
      node.version < 0 ||
      typeof node.revisionRef !== "string" ||
      !Number.isSafeInteger(node.revision) ||
      node.revision < 0 ||
      !["ACTIVE", "DELETED", "ARCHIVED"].includes(node.lifecycleState) ||
      !["", "PENDING", "SCANNING", "CLEAN", "QUARANTINED", "FAILED"].includes(
        node.scanState,
      ) ||
      !["", "ARTIFACT", "SKILL_BUNDLE", "MEMORY_RECORD"].includes(
        node.resourceKind,
      ) ||
      typeof node.selectable !== "boolean" ||
      ![
        "AVAILABLE",
        "DIRECTORY",
        "PERMISSION_REQUIRED",
        "IMMUTABLE_CONTEXT",
        "LIFECYCLE_BLOCKED",
        "ARTIFACT_USED_BY_SKILL",
        "ARTIFACT_NOT_ACTIVE",
        "ARTIFACT_NOT_DELETED",
        "ARTIFACT_HAS_BINDINGS",
        "ACTIVE_RUN_USES_ARTIFACT",
      ].includes(node.selectionReason) ||
      !Array.isArray(node.nextActions) ||
      node.nextActions.some(
        (action) =>
          ![
            "DOWNLOAD",
            "DELETE",
            "RESTORE",
            "PURGE",
            "ARCHIVE",
            "BIND",
          ].includes(action),
      ) ||
      (options.projectRef && options.projectRef !== node.projectRef) ||
      (!options.query.trim() && node.parentPath !== options.path) ||
      (options.query.trim() &&
        node.path !== options.path &&
        !node.path.startsWith(`${options.path.replace(/\/$/, "")}/`))
    )
      throw new Error("Invalid VFS node or scope");
    refs.add(node.ref);
  }
}

export function vfsEntityRoute(node: VfsNode): string | undefined {
  if (!node.projectRef || !node.entityRef) return undefined;
  const project = `/projects/${encodeURIComponent(node.projectRef)}`;
  const entity = encodeURIComponent(node.entityRef);
  switch (node.kind) {
    case "SKILL":
      return `${project}/context/skills/${entity}`;
    case "MEMORY":
      return `${project}/context/memory/${entity}`;
    case "PROJECT":
      return project;
    case "AGENT":
      return `${project}/agents/${entity}`;
    case "WORKFLOW":
      return `${project}/workflows/${entity}`;
    case "RUN":
      return `${project}/runs/${entity}`;
    case "AUTOMATION":
      return `${project}/automations?scheduleRef=${entity}`;
    case "ENVIRONMENT":
      return `${project}/environments/${entity}`;
    case "INPUT":
    case "RESULT":
    case "AVATAR":
      return `${project}/files?artifactRef=${entity}`;
    default:
      return undefined;
  }
}

File: services/internal/control-plane/internal/transport/grpc/vfs.go
package grpc

import (
	"context"

	controlplanev1 "github.com/codex-k8s/kodex/libs/go/controlplaneapi/gen/controlplane/v1"
	"github.com/codex-k8s/kodex/services/internal/control-plane/internal/domain/types/entity"
	"github.com/codex-k8s/kodex/services/internal/control-plane/internal/domain/types/query"
)

func (server *Server) ListVFSNodes(ctx context.Context, request *controlplanev1.ListVFSNodesRequest) (*controlplanev1.ListVFSNodesResponse, error) {
	p, err := principal(ctx, controlplanev1.PlatformQueryService_ListVFSNodes_FullMethodName)
	if err != nil {
		return nil, err
	}
	items, total, next, err := server.service.ListVFSNodes(ctx, p, query.Filter{ProjectRef: request.GetProjectRef(), ResourceRef: request.GetPath(), Query: request.GetQuery(), State: request.GetLifecycleState(), VFSKinds: vfsKinds(request.GetKinds()), Page: page(request.GetPage())})
	if err != nil {
		return nil, transportError(err)
	}
	response := &controlplanev1.ListVFSNodesResponse{Total: total, Page: &controlplanev1.PageInfo{NextPageToken: next}}
	for _, item := range items {
		response.Nodes = append(response.Nodes, castVFSNode(item))
	}
	return response, nil
}
func (server *Server) SearchVFS(ctx context.Context, request *controlplanev1.SearchVFSRequest) (*controlplanev1.SearchVFSResponse, error) {
	p, err := principal(ctx, controlplanev1.PlatformQueryService_SearchVFS_FullMethodName)
	if err != nil {
		return nil, err
	}
	items, total, next, err := server.service.SearchVFS(ctx, p, query.Filter{ProjectRef: request.GetProjectRef(), ResourceRef: request.GetPath(), Query: request.GetQuery(), State: request.GetLifecycleState(), VFSKinds: vfsKinds(request.GetKinds()), Page: page(request.GetPage())})
	if err != nil {
		return nil, transportError(err)
	}
	response := &controlplanev1.SearchVFSResponse{Total: total, Page: &controlplanev1.PageInfo{NextPageToken: next}}
	for _, item := range items {
		response.Nodes = append(response.Nodes, castVFSNode(item))
	}
	return response, nil
}
func castVFSNode(value entity.VFSNode) *controlplanev1.VFSNode {
	return &controlplanev1.VFSNode{Ref: value.Ref, Path: value.Path, ParentPath: value.ParentPath, Name: value.Name,
		Kind: controlplanev1.VFSNodeKind(controlplanev1.VFSNodeKind_value["VFS_NODE_KIND_"+value.Kind]), Directory: value.Directory,
		ProjectRef: value.ProjectRef, EntityRef: value.EntityRef, RunRef: value.RunRef, SizeBytes: value.SizeBytes,
		Digest: value.Digest, ModifiedAt: timestamp(value.ModifiedAt), Version: value.Version, RevisionRef: value.RevisionRef, Revision: value.Revision,
		LifecycleState: value.LifecycleState, ScanState: value.ScanState, ResourceKind: value.ResourceKind,
		Selectable: value.Selectable, SelectionReason: value.SelectionReason, NextActions: value.NextActions}
}

func vfsKinds(input []controlplanev1.VFSNodeKind) []string {
	result := make([]string, 0, len(input))
	for _, kind := range input {
		name := kind.String()
		const prefix = "VFS_NODE_KIND_"
		if len(name) > len(prefix) && name[:len(prefix)] == prefix {
			name = name[len(prefix):]
		}
		result = append(result, name)
	}
	return result
}
